pytorch_lightning/utilities/fetching.py

```diff
@@ -174,35 +208,183 @@ class DataFetcher(AbstractDataFetcher):
+            with self.fetching_context():
+                data = self.on_fetch_start()
+                with self.apply_profiler(f"fetch_next_{self.stage}_batch"):
+                    batch = next(self.dataloader_iter)
+                print(batch)
+                self.fetched += 1
+                self.on_fetch_end(batch, data)
+
+    def _consume_prefetched_batches(self) -> Generator:
+        self.done = True
+        while self.batches:
+            yield from self._yield_batch()
+
+    def _yield_batch(self) -> Generator:
+        self.wait()
+        batch = self.batches.pop(0)
+        is_last = len(self.batches) == 0
+        yield batch, is_last
+
+
+class InterBatchParallelismDataFetcher(DataFetcher):
+
+    """
+    This class implements `inter-batch-parallelism` algorithm which aims at hiding the latency of host-to-device copy
+    of input batches behind computational intensive operation.
+
+    Without parallization:
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:                          [HtoD][forward][backward]
+    batch 2:                                                   [HtoD][forward][backward]
+
+    With parallelization, the latency of HtoD copy can be hidden:
+
+    batch 0: [HtoD][forward][backward]
+    batch 1:       [HtoD]             [forward][backward]
+    batch 2:             [HtoD]                          [forward][backward]
+
+    """
+
+    def __init__(
+        self,
+        prefetch_batches: int = 0,
+    ) -> None:
+        super().__init__(prefetch_batches=prefetch_batches)
+
+        self.cuda_stream = torch.cuda.Stream()
+        self.events = []
+
+    @contextlib.contextmanager
+    def fetching_context(self):
+        """Wrap the batch fetching logic under a cuda stream"""
+        with torch.cuda.stream(self.cuda_stream):
+            yield
+
+    def on_fetch_start(self) -> "torch.cuda.Event":
+        # create a cuda event used to record the async stream of data to device.
+        return torch.cuda.Event()
+
+    def on_fetch_end(self, batch, event: torch.cuda.Event) -> None:
+        # move the batch to device and store it
+        super().on_fetch_end(batch)
+
+        # record event and store the event
+        event.record()
+        self.events.append(event)
+
+    def wait(self) -> None:
+        # pop first event from the queue and wait for the batch to be available on device.
+        event = self.events.pop(0)
+        event.wait()
+
+
+class TrainingStepDataLoaderIter:
+
+    """This class is a wrapper to keep track of dataloader iterator fetching event while left entirely to user control."""
+
+    def __init__(self, iterator: Iterator, data_fetcher: "AbstractDataFetcher"):
+        self.iterator = iterator
+        self.data_fetcher = data_fetcher
+
+    def __iter__(self) -> "TrainingStepDataLoaderIter":
+        return self
+
+    def __next__(self) -> Any:
+        self.data_fetcher.fetched += 1
+        try:
+            return next(self.iterator)
```

Comment:
does this fetch need to be in the try/catch, after the `next(self.iterator)` happens? what is fetched supposed to measure? the # of times next is called, or the number of iterations sucessfully run?

Reply:
This is done for Lightning to track iterator exhaustion while being fetched by the users.